Smart union: add active_name() and callWithActive()

gavin-norman-sociomantic opened the pull request.

Commit:
Don't mark unittest as usage example

Commit:
Move usage example to documented unittest

(And fix compilation error ;)

Commit:
Add missing template params doc

Commit:
Extend unittest to check active member after setting

nemanja-boric-sociomantic commented:
You seems to have D2 failures: https://travis-ci.org/sociomantic-tsunami/ocean/builds/209681331

```
./src/ocean/core/SmartUnion.d(252): Error: function ocean.core.SmartUnion.callWithActive!(print, SmartUnion!(TestUnion)).callWithActive cannot access frame of function ocean.core.SmartUnion.__unittestL259_270.printActiveUnionField
./src/ocean/core/SmartUnion.d(252): Error: function ocean.core.SmartUnion.callWithActive!(print, SmartUnion!(TestUnion)).callWithActive cannot access frame of function ocean.core.SmartUnion.__unittestL259_270.printActiveUnionField
./src/ocean/core/SmartUnion.d(283): Error: template instance ocean.core.SmartUnion.callWithActive!(print, SmartUnion!(TestUnion)) error instantiating
make: *** [/home/travis/build/sociomantic-tsunami/ocean/build-d2/production/tmp/allunittests] Error 1
```

gavin-norman-sociomantic commented:
Updated.

david-eckardt-sociomantic commented on src/ocean/core/SmartUnion.d:
```diff
@@ -254,6 +289,33 @@ private template MemberList ( uint i, size_t len, U )
+        const MemberStringList = `,"` ~ FieldName!(i, U) ~ `"` ~
+            MemberStringList!(i + 1, U);
+    }
+}
```
Suggest generating the actual array of strings here like
```d
// Template code would probably be cumbersome so use CTFE
private static istring[] memberStringList ( U ) ( )
{
    istring[] member_strings = ["none"[]];
    foreach (i, Field; typeof(U.tupleof))
        member_strings ~= FieldName!(i, U);
    return member_strings;
}

// inside SmartUnionIntern
istring[] active_names = memberStringList!(U)();
```
Btw. why did you put `MemberStringList` outside the `SmartUnion` and `SmartUnionIntern` struct?

david-eckardt-sociomantic replied:
> Btw. why did you put `MemberStringList` outside the `SmartUnion` and `SmartUnionIntern` struct?

I see why: `MemberList` is outside as well. I didn't read this code for a long time ;)
Still you *could* put `MemberStringList` inside `SmartUnion/Intern`.

gavin-norman-sociomantic replied:
Regarding both points (the location of the template and using a template vs CTFE), I simply wrote code in the same style as the existing code in this module. I agree that the CTFE solution looks a lot nicer, though, so will change that.

david-eckardt-sociomantic commented on src/ocean/core/SmartUnion.d:
```diff
@@ -212,6 +212,86 @@ version (UnitTest)
+{
+    alias TemplateInstanceArgs!(SmartUnion, SU) SUArgs;
+    static assert(SUArgs.length == 1);
+    alias SUArgs[0] U;
```
Hm, why didn't you put this method in `SmartUnion` so you don't have to derive the union type? Btw. I would probably use a simple `alias typeof(smart_union._) U`.

david-eckardt-sociomantic replied:
> Btw. I would probably use a simple `alias typeof(smart_union._) U`.

Never mind, this may not work.

david-eckardt-sociomantic replied:
> Btw. I would probably use a simple `alias typeof(smart_union._) U`.

`alias typeof(smart_union._.u) U`, of course.

gavin-norman-sociomantic replied:
>Hm, why didn't you put this method in SmartUnion so you don't have to derive the union type?

```
    Note: declared at module-scope (rather than nested inside the SmartUnion
    template) to work around limitations of template alias parameters. (Doing it
    like this allows it to be called with a local name.)
```

david-eckardt-sociomantic commented on src/ocean/core/SmartUnion.d:
```diff
@@ -212,6 +212,86 @@ version (UnitTest)
+    if ( smart_union._.active == smart_union.Active.none )
+        return;
+
+    assert(smart_union._.active >= 1);
```
`smart_union.Active.none` is intentionally defined to be zero to allow for
```d
if ( !smart_union._.active ) return;
```
in situations like this, which makes the assertion unnecessary (at least for me ;) ).

gavin-norman-sociomantic replied:
I see, yeah that's a bit cleaner.

Commit:
Add SmartUnion : callWithActive

Allows a user-specified function to be called with the currently active
field of a smart-union.

gavin-norman-sociomantic commented:
Updated:
* CTFE names list.
* `alias typeof(smart_union._.u) U`

mathias-lang-sociomantic commented on src/ocean/core/SmartUnion.d:
```diff
@@ -220,6 +330,34 @@ private struct SmartUnionIntern ( U )
+
+    ***************************************************************************/
+
+    istring[] active_names = member_string_list();
```
`const` ?

gavin-norman-sociomantic replied:
Updated.

mathias-lang-sociomantic commented on src/ocean/core/SmartUnion.d:
```diff
@@ -220,6 +330,34 @@ private struct SmartUnionIntern ( U )
+
+    ***************************************************************************/
+
+    static private istring[] member_string_list ( )
```
`private static Immut!(istring[])` so you can assign it at CT ?

david-eckardt-sociomantic replied:
:+1: but I think it is actually assigned at CT already.

mathias-lang-sociomantic replied:
No it's not, because it isn't assigned to a manifest constant (` istring[] active_names = member_string_list();`).

mihails-strasuns-sociomantic replied:
All initializers in D are required to be CTFE-evaluated, not matter if LHS is constant or not. Also it is allowed to implicitly convert any kind of mutability qualifier from RHS to the one required by LHS upon initialization for that exact reason.

david-eckardt-sociomantic commented on src/ocean/core/SmartUnion.d:
```diff
@@ -179,6 +212,83 @@ version (UnitTest)
+
+public void callWithActive ( alias Callable, SU ) ( SU smart_union )
+{
+    static assert(is(TemplateInstanceArgs!(SmartUnion, SU)));
```
You like to go on number secure :)

gavin-norman-sociomantic replied:
I don't want people passing any old crap to my function!

david-eckardt-sociomantic replied:
Funktionstürsteher

Commit:
Add SmartUnion.active_name

nemanja-boric-sociomantic commented on src/ocean/core/SmartUnion.d:
```diff
@@ -103,6 +95,31 @@ struct SmartUnion ( U )
 {
+    union MyUnion
+    {
+        int    x;
```
nitpick: indentation here is screwed.

gavin-norman-sociomantic replied:
Ha oh yeah. Updated.

nemanja-boric-sociomantic commented on src/ocean/core/SmartUnion.d:
```diff
@@ -226,6 +253,34 @@ private struct SmartUnionIntern ( U )
+
+    static private istring[] member_string_list ( )
+    {
+        istring[] names = ["none"[]];
```
I wonder if the empty string would be a better choice here (if the user is allowed to have the `none` as a member)?

gavin-norman-sociomantic replied:
The smart-union enum member is actually called `none`, so this is entirely accurate. https://github.com/gavin-norman-sociomantic/ocean/blob/52961089d40181042c1680ad7af43166c4d34861/src/ocean/core/SmartUnion.d#L247

nemanja-boric-sociomantic replied:
Sorry, I've missed this!

nemanja-boric-sociomantic commented on src/ocean/core/SmartUnion.d:
```diff
@@ -212,6 +212,83 @@ version (UnitTest)
+    // This example is D2 only because it requires a function template,
+    // `print`, and D1 doesn't allow defining a function template at the scope
+    // of a function, including a `unittest`. In D1 this example works if
+    // `print`, is defined outside of function scope.
```
"outside of `unittest` scope"?

gavin-norman-sociomantic replied:
Hm, it's already mentioned that function scope = unittest scope.

nemanja-boric-sociomantic replied:
Yeah, true, it's just noise from me.

nemanja-boric-sociomantic reviewed (approved):
LGTM

nemanja-boric-sociomantic commented:
Maybe @david-eckardt-sociomantic wants to give it a final look?
